pkg/component/gardenerapiserver/secrets.go

```diff
@@ -0,0 +1,37 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gardenerapiserver
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+func (g *gardenerAPIServer) reconcileSecretServer(ctx context.Context) (*corev1.Secret, error) {
+	return g.secretsManager.Generate(ctx, &secretsutils.CertificateSecretConfig{
+		Name:                        secretNameServer,
+		CommonName:                  DeploymentName,
+		DNSNames:                    append(kubernetesutils.DNSNamesForService(DeploymentName, g.namespace), kubernetesutils.DNSNamesForService(DeploymentName, metav1.NamespaceSystem)...),
```

Comment:
Why do we add `gardener-apiserver.kube-system.svc.cluster.local` (and shorter variants) as DNS name? Do we expect gardener apiserver to have a split personality running in both the specified namespace, but also in `kube-system` ;-) ?

Reply:
🤣 
The `virtual-garden-kube-apiserver` talks to `gardener-apiserver` via `gardener-apiserver.kube-system` since the `APIService`s in the virtual cluster point to this `Service`. However, I still wanted to make it possible for other components in the runtime cluster to directly talk to `gardener-apiserver` via the `Service` in the `garden` namespace, hence via `gardener-apiserver.garden`.
Most components will use the "default" path via KAPI to talk to GAPI, though.

Reply:
Ok, that explains it.
Why do we deploy the `APIService` into `kube-system` though? Would it not make sense to have it next the actual workload, i.e. in `garden` namespace?

Reply:
The `APIService` is a global resource (not namespaced). Also, it's deployed to the virtual cluster (where no workload is at all).

Reply:
It is about the virtual cluster again :-( . I thought about `kube-system` in the runtime cluster...